crates/turborepo-lib/src/shim.rs

```diff
@@ -461,129 +460,62 @@ impl InferInfo {
-                    root: current.path.clone(),
-                    mode: RepoMode::MultiPackage,
-                });
-
-            // More than one potential root.
-            // Closest is not RepoMode::MultiPackage
-            // We attempt to prove that the closest is a workspace of a parent.
-            // Failing that we just choose the closest.
-            } else {
-                for ancestor_infer in check_roots {
-                    if ancestor_infer.is_workspace_root_of(&current.path) {
-                        return Ok(Self {
-                            root: ancestor_infer.path.clone(),
-                            mode: RepoMode::MultiPackage,
-                        });
-                    }
-                }
-
-                // We have eliminated RepoMode::MultiPackage as an option.
-                // We must exhaustively check before this becomes the answer.
-                return Ok(Self {
-                    root: current.path.clone(),
-                    mode: RepoMode::SinglePackage,
-                });
-            }
+impl From<InferInfo> for RepoState {
+    fn from(root: InferInfo) -> Self {
+        Self {
+            mode: root.repo_mode(),
+            root: root.path,
         }
-
-        // If we're here we didn't find a valid root.
-        Err(anyhow!("Root could not be inferred."))
     }
+}
 
-    /// Infers `RepoState` from current directory.
+impl RepoState {
+    /// Infers `RepoState` from a reference path
     ///
     /// # Arguments
     ///
-    /// * `current_dir`: Current working directory
+    /// * `reference_dir`: Turbo's invocation directory
     ///
     /// returns: Result<RepoState, Error>
     pub fn infer(reference_dir: &AbsoluteSystemPath) -> Result<Self> {
-        let potential_turbo_roots = RepoState::generate_potential_turbo_roots(reference_dir);
-        RepoState::process_potential_turbo_roots(potential_turbo_roots)
+        reference_dir
+            .ancestors()
+            .filter_map(|path| {
+                PackageJson::load(&path.join_component("package.json"))
+                    .ok()
+                    .map(|package_json| {
+                        // FIXME: We should save this package manager that we detected
+                        let workspace_globs =
+                            PackageManager::get_package_manager(path, Some(&package_json))
+                                .and_then(|mgr| mgr.get_workspace_globs(path))
+                                .ok();
+
+                        InferInfo {
+                            path: path.to_owned(),
+                            workspace_globs,
+                        }
+                    })
+            })
+            .reduce(|current, candidate| {
+                if current.repo_mode() == RepoMode::MultiPackage {
+                    // We already have a multi-package root, go with that
+                    current
+                } else if candidate.is_workspace_root_of(&current.path) {
```

Comment:
terminology question: what does "workspace root" in `is_workspace_root_of` mean?

(cc @mrmckeb @codybrouwers who are interested in defining our vocabulary also)